### PR TITLE
stdenv: fix succeedOnFailure

### LINKED ISSUE #200663: `succeedOnFailure = true` broken on master
### Describe the bug
`succeedOnFailure` is a feature of nixpkgs stdenv and hydra that makes it possible to host the build's output on hydra even in the light of a failure by respecting a file `$out/nix-support/failed`.
This feature is apparently broken in `master`, whereas it used to work in 21.05.

### Steps To Reproduce

Put the following file into the `nixpkgs` root and build it:
```nix
let
  pkgs = import ./. { };
in pkgs.runCommand "foo" {
  succeedOnFailure = true;
  } ''
    mkdir $out
    echo foo > $out/foo
    exit 1
  ''
```

### Expected behavior
The build should succeed with a similar output

```
build failed with exit code 1 (ignored)
/nix/store/966yi36fc32g9w1zwr95b6qh3xfh047a-foo
```

instead the build fails

```
building '/nix/store/akiv0nhwijmjmzf4ja92qxq955bqj66r-foo.drv'...
build failed with exit code 1 (ignored)
error: builder for '/nix/store/akiv0nhwijmjmzf4ja92qxq955bqj66r-foo.drv' failed with exit code 1;
       last 1 log lines:
       > build failed with exit code 1 (ignored)
       For full logs, run 'nix log /nix/store/akiv0nhwijmjmzf4ja92qxq955bqj66r-foo.drv'.
```

### Notify maintainers
@SuperSandro2000 as bisecting this issue did point to https://github.com/NixOS/nixpkgs/pull/161079 as the first bad commit.

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Description of changes

Closes #200663

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
I believe this effect is better achieved via https://github.com/NixOS/nixpkgs/pull/196251
Otherwise lgtm.

**Review by knedlsepp — APPROVED** [review]
Used the following to confirm that it works as intended:

```
let
  pkgs = import ./. { };
in pkgs.runCommand "foo" {
  succeedOnFailure = true;
  } ''
    mkdir $out
    echo foo > $out/foo
    exit 1
  ''
```

Thank you! :-)

**Comment by roberth** [comment]
Apparently ofborg is quick enough to test a stdenv rebuild, so I've added @knedlsepp's test to the `tests` attribute.

**Comment by knedlsepp** [comment]
@roberth thanks for adding a test. One note: Due to the nature of `succeedOnFailure` this test might show up on hydra as failed, which might be confusing. I'm not quite sure how to address that though.

**Comment by roberth** [comment]
@knedlsepp as far as I know, `succeedOnFailure` is just a stdenv concept that does not need anything beyond a normal derivation build to work correctly. Nix does not have a special interpretation of it, and I don't expect Hydra to have it either. (If it does, it will just succeed on Hydra, but still be useful on ofborg and all the other tools)

It does not invert the exit status either, like https://github.com/NixOS/nixpkgs/pull/196251 would.

**Review by roberth — APPROVED** [review]

**Comment by knedlsepp** [comment]
@roberth 

> Nix does not have a special interpretation of it, and I don't expect Hydra to have it either.

It's true that nix itself does not have any special interpretation, but hydra does. It renders these outputs with a red cross as I'd they failed, but offers downloads in `nix-support/hydra-build-products`.
It is used on release-tools in nixpkgs sich as coverageReport.


https://github.com/NixOS/hydra/commit/039d1bed21e6fae4be0d3b532703e084d4a0f23d

Anyway, I couldn't find the test that you added in the most recent hydra evaluation, so it doesn't really matter. 😄

**Comment by roberth** [comment]
That's a bad hack for something that should be solved by Nix itself, but alright then. By default the package `tests` attribute doesn't get built on Hydra, so it's probably fine.

File: pkgs/stdenv/generic/default.nix [hunk 1]
```diff
@@ -175,6 +175,11 @@ let
       # without running any commands. Because this will also skip `shopt -s extglob`
       # commands and extglob affects the Bash parser, we enable extglob always.
       shellDryRun = "${stdenv.shell} -n -O extglob";
+
+      tests = {
+        succeedOnFailure = import ../tests/succeedOnFailure.nix { inherit stdenv; };
+      };
+      passthru.tests = lib.warn "Use `stdenv.tests` instead. `passthru` is a `mkDerivation` detail." stdenv.tests;
```
Comment (SuperSandro2000) [inline pkgs/stdenv/generic/default.nix:182]: Should this be triggered by ``nix eval``? I have encountered this on my system and other peoples and I am unsure if the issue is here or in nix eval.

**Pull request merged**